docs/nala.8.md
---
title: Nala
section: 8
header: User Manual
footer: Nala 0.9.1
date: 09 June 2022
---
# NAME
Nala - A prettier front-end for libapt-pkg

# SYNOPSIS
**nala** <*command*> [*\--options*]...

# DESCRIPTION
**nala** is a prettier front-end for libapt-pkg. The goals of **nala** are to be similar to **apt** but add quality of life changes, and improve the output to make it more reader friendly.

**install**
: > **install** works similar to the way it does in **apt**. **nala** takes multiple packages as arguments and will install all of them just like **apt**.

: > **nala** downloads packages in parallel. Along with this **nala** can download packages from multiple mirrors concurrently to speed up package downloads further. We put in place a limit to 2 packages per mirror so we don't put too much pressure on them. This caps out at a maximum of 16.

	*-f, \--fix-broken* attempts to fix broken packages. By default Nala will attempt to fix them. This switch exists so if you install something with **dpkg** and it needs extra dependencies to work you can run **nala install --fix-broken**.

	*\--no-fix-broken* will stop **nala** from doing some extra checks during an install. Using this can result in a broken install.

**remove**
: > **remove** works similar to the way it does in **apt**. Our noticeable differences here include improved output on what will be removed, no need for running an autoremove, **nala** will handle that for you.

**purge**
: > **purge** works similar to **remove**, but also gets rid of configuration files.

**update**
: > **update** updates the package list. Works exactly like **apt update** with better formatting.

**upgrade**
: > **upgrade** upgrades packages installed on the system.

	By default **nala** will run the equivalent of **apt update && apt full-upgrade --auto-remove**. If you are just looking to update the package cache and not actually perform an upgrade you can use **nala** *\--update*.

	*\--full* / *\--no-full* Toggle full-upgrade. *\--no-full* will run a regular upgrade which won't remove or install packages. By default **nala** uses a *full-upgrade*.

	*\--exclude* packages to exclude when updating. Glob enabled, for example. **nala update** *\--exclude* *linux-image**

**autoremove**, **autopurge**
: > **autoremove/autopurge** removes or purges packages that are no longer needed.

	*\--config* Purge packages that are not installed and have configuration files on the system. Must be used with *\--purge* or **autopurge**

**clean**
: > **clean** is exactly like running *apt clean*.

	*\--fetch* Remove `nala-sources.list` generated from **nala fetch**

	*\--lists* Remove package lists located in `/var/lib/apt/lists/`

**fetch**
: > **fetch** is our first command that doesn't have an **apt** counterpart.

	**nala** will parse either the **Debian** mirror list from *https://mirror-master.debian.org/status/Mirrors.masterlist*, or the **Ubuntu** mirror list from *https://launchpad.net/ubuntu/+archivemirrors-rss* and then fetch (3 by default) mirrors that we have determined are the closest to you.

	**nala** will attempt to detect your distro and release by default. Don't worry if it's not able too, as you can specify it manually with some switches below.

	This functionality is much like you would expect from **netselect** and **netselect-apt**.

	*\--debian* lets you specify that you're using **Debian** and what release you're using. *\--debian sid*

	*\--ubuntu* This is just the **Ubuntu** version of the switch above. *\--ubuntu jammy*

	*\--country* lets you specify your *country* when fetching mirrors. Use the ISO country code. You don't have to use this as we test latency anyway, but it can speed things up as there are less mirrors to test.

	*\--foss* on *Debian* will ensure that you don't get the *contrib* or *non-free* repos. Using this on *Ubuntu* does nothing.

	*\--sources* add the source repos for the mirrors if it exists.

	*\--count* choose the number of viable mirrors to display. by default it's 16, but if you want more you got it.

	*\--auto* run fetch uninteractively. Will still prompt for overwrite.

	*\--fetches* lets you determine the amount of mirrors to fetch between 1-10. 3 is the default. This switch only works with *\--auto*

**show**
: > **show** works exactly like the **apt** version except our output is a little easier to read. **show** will accept multiple packages as arguments.

	*\-a, \--all-versions* will print information about each version that is currently in the cache.

**search**
: > **search** all packages in the cache with a regex string. By default searches package names, source names, and descriptions.

	*\-n, \--names* will search only package names. This is much faster than searching everything.

	*\-i, \--installed* will search only installed packages.

	*\-N, \--nala-installed* Only packages explicitly installed with Nala.

	*\-u, \--upgradable* will search only upgradable packages.

	*\-a, \--all-versions* will print information about each version that is currently in the cache.

	*\-A, \--all-arches* will search all architectures of a package.

	*\-V, \--virtual* will search only virtual packages.

	*\--full* will print the full description of each package found rather than just the short description.

**list**
: > **list** packages in the cache.

	The switches here are the same as the **search** command except *\-n, \--names* and *\-A, \--all-arches* are absent.

**history**
: > **history** is our other new command. Every **install**, **remove**, or **upgrade** command is stored with an id. You can use **history** to view these in a summary style view, and even go more in depth with **history info [id]**. If you're familiar with how *Fedora's* **dnf history** command works, then you'll feel right at home. That's what we drew inspiration from.

	**history info [id]** prints a more detailed summary of what happened in that transaction.

	**history undo [id]** will undo the transaction.

	**history redo [id]** will redo the transaction.

	Currently only **remove** and **install** commands are available for **undo** and **redo**

	**history clear [id]** will remove the transaction from the history. Alternatively you can use *\--all* which will remove the entire history

	*\--installed* prints the names of packages explicitly installed with Nala.

# OPTIONS
**\--help**
: >*\--help* will print out a help message for each subcommand. **nala install** *\--help* is a different message than **nala update** *\--help*.

**-y, \--assume-yes** / **-n, \--assume-no**
: >*\--assume-yes* will automatically select yes for any prompts which may need your input. This can potentially be dangerous.

	*\--assume-no* will automatically select no for any prompts which may need your input.

**-d, \--download-only**
: >*\--download-only* will do just that, download packages only. It will not unpack or configure anything.

**-v, \--verbose**
: >*\--verbose* prints more information that sticks around. By default we use "Progress Bars" to hide some things we consider unnecessary. Using *\--verbose* removes the progress bars and prints that information.

**-o, \--option**
: >*\--option* can pass any configuration option to **apt**/**dpkg** like Dpkg::Options::="--force-confnew" or APT::Default-Release="stable".

**\--purge**
: >*\--purge* any packages that would removed during the transaction.

**\--fix-broken** / **\--no-fix-broken**
: >*\--fix-broken* / *\--no-fix-broken/* toggles for fixing broken packages

	*\--fix-broken* will attempt to fix broken packages. **nala install** *\--fix-broken*

	*\--no-fix-broken* can be used if you don't want Nala to attempt to fix broken packages.

**\--install-recommends** / **\--no-install-recommends**
: >*\--install-recommends* / *\--no-install-recommends* toggles for recommended packages

	*\--install-recommends** sets recommended packages to be installed. [Default]

	*\--no-install-recommends* stops the installation of recommended packages. If this option is selected Nala will display the recommended packages that won't be installed.

**\--install-suggests** / **\--no-install-suggests**
: >*\--install-suggests* / *\--no-install-suggests* toggles for suggested packages

	*\--install-suggests* installs suggested package. If this option is not selected Nala will display the suggested packages that won't be installed.

	*\--no-install-suggests* suggested packages will not be installed. [Default]

**\--update** / **\--no-update**
: >*\--update* / *\--no-update* toggles for updating the package list

	*\--no-update* skips updating the package cache if for whatever reason you would like to skip that.

	*\--update* updates the package cache. Typically run as **nala install** *\--update* or **nala** *\--update*.

**\--autoremove** / **\--no-autoremove**
: >*\--autoremove* / *\--no-autoremove* toggles for autoremoving packages

	*\--autoremove* nala will autoremove unneeded packages. [Default]

	*\--no-autoremove* stops nala from autoremoving packages. Packages that need to be removed in order to continue the upgrade will still be removed.

**\--remove-essential**
: >*\--remove-essential* allows the removal of essential packages. This is very dangerous, but we thought you should have the option.

**\--debug**
: >*\--debug* prints helpful information for solving issues. If you're submitting a bug report try running the command again with *\--debug* and providing the output to the devs, it will be helpful.

**\--version**
: >*\--version* prints the version of nala you have installed and exits.

**\--license**
: >*\--license* reads the GPLv3 which **nala** is licensed under.

**\--raw-dpkg**
: >*\--raw-dpkg* forces **nala** not to mess with **dpkg** while running operations. This disables all formatting and it would look as if you were using **apt**. A more indepth explanation for what this switch does, **nala** will fork a tty instead of a pty for **dpkg**.

	**nala** will also not display a progress bar for **dpkg** with this turned on. Additionally the language of the output will not be forced into English for this mode.

# EXAMPLES
**nala install** *\--update* **wine**
: >downloads and installs wine, updating the package cache first.

**nala upgrade** *\--no-update*
: >upgrades the system without updating the package cache.

**nala fetch** *\--fetches* **8** *\--country* **US**
: >fetches 8 mirrors in the United states to max out parallel downloads.

# AUTHORS
Blake Lee <*https://gitlab.com/volitank*> <*blake@volian.org*>

volian-team <*https://gitlab.com/volian*> <*volian-devel@volian.org*>

# BUGS
Submit bug reports online at: <*https://gitlab.com/volian/nala/-/issues*>

# SEE ALSO
Sources at: <*https://gitlab.com/volian/nala*>


debian/NEWS
nala (0.9.0) unstable; urgency=medium
  The configuration file for Nala at `/etc/nala/nala.conf`
  has been changed to toml. This change is BREAKING as the older config
  file will no longer work. Nala will still run with the old config,
  but there will be a syntax error and it will use the program defaults.

 -- Blake Lee <blake@volian.org>  Mon, 23 May 2022 15:53:49 -0400


nala/nala.py
#                 __
#    ____ _____  |  | _____
#   /    \\__  \ |  | \__  \
#  |   |  \/ __ \|  |__/ __ \_
#  |___|  (____  /____(____  /
#       \/     \/          \/
#
# Copyright (C) 2021, 2022 Blake Lee
#
# This file is part of nala
# nala is based upon apt-metalink https://github.com/tatsuhiro-t/apt-metalink
#
# nala is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# nala is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with nala.  If not, see <https://www.gnu.org/licenses/>.
"""Main module for Nala which facilitates apt."""
from __future__ import annotations

import re
import sys
from subprocess import run
from typing import Generator, Optional

import apt_pkg
import typer
from apt.package import Package, Version

from nala import _, color
from nala.cache import Cache
from nala.constants import (
	ARCHIVE_DIR,
	CAT,
	DPKG_STATE,
	ERROR_PREFIX,
	LISTS_DIR,
	LISTS_PARTIAL_DIR,
	NALA_SOURCES,
	PARTIAL_DIR,
	PKGCACHE,
	SRCPKGCACHE,
)
from nala.error import BrokenError, pkg_error
from nala.history import get_history, get_list
from nala.install import (
	auto_remover,
	check_broken,
	check_state,
	check_term_ask,
	fix_excluded,
	get_changes,
	install_local,
	package_manager,
	setup_cache,
	split_local,
)
from nala.options import (
	ALL_ARCHES,
	ALL_VERSIONS,
	ASSUME_YES,
	AUTO_REMOVE,
	CONFIG,
	DEBUG,
	DOWNLOAD_ONLY,
	FETCH,
	FIX_BROKEN,
	FULL,
	INSTALLED,
	LISTS,
	MAN_HELP,
	NALA_INSTALLED,
	NAMES,
	OPTION,
	PURGE,
	RAW_DPKG,
	RECOMMENDS,
	REMOVE_ESSENTIAL,
	SUGGESTS,
	UPDATE,
	UPGRADABLE,
	UPGRADEABLE,
	VERBOSE,
	VIRTUAL,
	arguments,
	nala,
)
from nala.rich import ELLIPSIS
from nala.search import iter_search, search_name
from nala.show import additional_notice, pkg_not_found, show_main
from nala.utils import (
	PackageHandler,
	ask,
	command_help,
	dedupe_list,
	eprint,
	get_version,
	iter_remove,
	sudo_check,
	vprint,
)

nala_pkgs = PackageHandler()


def _fix_broken(nested_cache: Cache | None = None) -> None:
	"""Attempt to fix broken packages, if any."""
	cache = nested_cache or setup_cache()
	print("Fixing Broken Packages...")
	cache.fix_broken()

	if nested_cache:
		print(color(_("There are broken packages that need to be fixed!"), "YELLOW"))
		print(
			_("You can use {switch} if you'd like to try without fixing them.").format(
				switch=color("--no-fix-broken", "YELLOW")
			)
		)
	else:
		check_state(cache, nala_pkgs)
	get_changes(cache, nala_pkgs, "fix-broken")


def _remove(pkg_names: list[str]) -> None:
	sudo_check()

	cache = setup_cache()
	check_state(cache, nala_pkgs)

	pkg_names = cache.glob_filter(dedupe_list(pkg_names))
	pkg_names = cache.virtual_filter(pkg_names, remove=True)
	broken, not_found, ver_failed = check_broken(
		pkg_names,
		cache,
		remove=True,
	)

	for pkg_name in not_found[:]:
		if cache.is_any_virtual(pkg_name):
			not_found.remove(pkg_name)
			pkg_names.remove(pkg_name)

	if not_found or ver_failed:
		pkg_error(not_found, cache)

	nala_pkgs.user_explicit = [cache[pkg_name] for pkg_name in pkg_names]
	if not package_manager(pkg_names, cache, remove=True):
		BrokenError(cache).broken_remove(broken)

	auto_remover(cache, nala_pkgs)
	get_changes(cache, nala_pkgs, "remove")


def _install(pkg_names: list[str] | None, ctx: typer.Context) -> None:
	sudo_check(pkg_names)
	if not pkg_names:
		if arguments.fix_broken:
			_fix_broken()
			return
		ctx.fail(_("{error} Missing packages to install").format(error=ERROR_PREFIX))

	pkg_names = dedupe_list(pkg_names)  # type: ignore[arg-type]
	check_state(cache := setup_cache(), nala_pkgs)
	not_exist = split_local(pkg_names, cache, nala_pkgs.local_debs)
	install_local(nala_pkgs, cache)

	pkg_names = cache.glob_filter(pkg_names)
	pkg_names = cache.virtual_filter(pkg_names)
	broken, not_found, ver_failed = check_broken(pkg_names, cache)
	not_found.extend(not_exist)

	if not_found or ver_failed:
		pkg_error(not_found, cache)

	nala_pkgs.user_explicit = [cache[pkg_name] for pkg_name in pkg_names]
	if not package_manager(pkg_names, cache):
		if not (error := BrokenError(cache, broken)).broken_install():
			error.unmarked_error(nala_pkgs.user_explicit)

	auto_remover(cache, nala_pkgs)
	get_changes(cache, nala_pkgs, "install")


def remove_completion(ctx: typer.Context) -> Generator[str, None, None]:
	"""Complete remove command arguments."""
	if not DPKG_STATE.exists():
		return
	regex = r"ok installed|half-installed|unpacked|half-configured"
	if "purge" in ctx.command_path:
		regex += r"|config-files"

	status = re.compile(regex)
	for package in DPKG_STATE.read_text(encoding="utf-8").split("\n\n"):
		pkg_name = pkg_status = None
		for line in package.splitlines():
			if len(feilds := line.split(": ")) == 1:
				continue
			if "Package" in feilds:
				pkg_name = feilds[1]
			if "Status" in feilds and status.findall(feilds[1]):
				pkg_status = feilds[1]
		if pkg_name and pkg_status:
			yield pkg_name


def package_completion(cur: str) -> Generator[str, None, None]:
	"""Complete install command arguments."""
	yield from run(
		["apt-cache", "--no-generate", "pkgnames", cur]
		if PKGCACHE.exists()
		else ["apt-cache", "pkgnames", cur],
		capture_output=True,
		check=True,
		text=True,
	).stdout.split()


@nala.command("update", help=_("Update package list."))
# pylint: disable=unused-argument
def _update(
	debug: bool = DEBUG,
	raw_dpkg: bool = RAW_DPKG,
	dpkg_option: list[str] = OPTION,
	verbose: bool = VERBOSE,
	man_help: bool = MAN_HELP,
) -> None:
	"""Update package list."""
	sudo_check()
	arguments.update = True
	setup_cache().print_upgradable()


@nala.command(help=_("Update package list and upgrade the system."))
# pylint: disable=unused-argument,too-many-arguments, too-many-locals
def upgrade(
	exclude: Optional[list[str]] = typer.Option(
		None,
		metavar="PKG",
		help=_("Specify packages to exclude during upgrade. Accepts glob*"),
	),
	purge: bool = PURGE,
	debug: bool = DEBUG,
	raw_dpkg: bool = RAW_DPKG,
	download_only: bool = DOWNLOAD_ONLY,
	remove_essential: bool = REMOVE_ESSENTIAL,
	full: bool = typer.Option(True, help=_("Toggle full-upgrade")),
	update: bool = UPDATE,
	auto_remove: bool = AUTO_REMOVE,
	install_recommends: bool = RECOMMENDS,
	install_suggests: bool = SUGGESTS,
	fix_broken: bool = FIX_BROKEN,
	assume_yes: bool = ASSUME_YES,
	dpkg_option: list[str] = OPTION,
	verbose: bool = VERBOSE,
	man_help: bool = MAN_HELP,
) -> None:
	"""Update package list and upgrade the system."""
	sudo_check()

	def _upgrade(
		full: bool = True,
		exclude: list[str] | None = None,
		nested_cache: Cache | None = None,
	) -> None:
		"""Upgrade pkg[s]."""
		cache = nested_cache or setup_cache()
		check_state(cache, nala_pkgs)

		is_upgrade = tuple(cache.upgradable_pkgs())
		protected = cache.protect_upgrade_pkgs(exclude)
		try:
			cache.upgrade(dist_upgrade=full)
		except apt_pkg.Error:
			if exclude:
				exclude = fix_excluded(protected, is_upgrade)
				if ask(_("Would you like us to protect these and try again?")):
					cache.clear()
					_upgrade(full, exclude, cache)
					sys.exit()
				sys.exit(
					_("{error} You have held broken packages").format(
						error=ERROR_PREFIX
					)
				)

			BrokenError(
				cache, tuple(pkg for pkg in cache if pkg.is_inst_broken)
			).broken_install()

		if kept_back := tuple(
			pkg for pkg in is_upgrade if not (pkg.marked_upgrade or pkg.marked_delete)
		):
			BrokenError(cache, kept_back).held_pkgs()
			check_term_ask()
			cache.upgrade(dist_upgrade=full)

		auto_remover(cache, nala_pkgs)
		get_changes(cache, nala_pkgs, "upgrade")

	_upgrade(full, exclude)


@nala.command(help=_("Install packages."))
# pylint: disable=unused-argument,too-many-arguments,too-many-locals
def install(
	ctx: typer.Context,
	pkg_names: Optional[list[str]] = typer.Argument(
		None,
		metavar="PKGS ...",
		help=_("Package(s) to install"),
		autocompletion=package_completion,
	),
	purge: bool = PURGE,
	debug: bool = DEBUG,
	raw_dpkg: bool = RAW_DPKG,
	download_only: bool = DOWNLOAD_ONLY,
	remove_essential: bool = REMOVE_ESSENTIAL,
	update: bool = UPDATE,
	auto_remove: bool = AUTO_REMOVE,
	install_recommends: bool = RECOMMENDS,
	install_suggests: bool = SUGGESTS,
	fix_broken: bool = FIX_BROKEN,
	assume_yes: bool = ASSUME_YES,
	dpkg_option: list[str] = OPTION,
	verbose: bool = VERBOSE,
	man_help: bool = MAN_HELP,
) -> None:
	"""Install packages."""
	_install(pkg_names, ctx)


@nala.command(help=_("Remove packages."))
@nala.command("purge", help=_("Purge packages."))
@nala.command("uninstall", hidden=True)
# pylint: disable=unused-argument,too-many-arguments
def remove(
	pkg_names: list[str] = typer.Argument(
		...,
		metavar="PKGS ...",
		help=_("Package(s) to remove/purge"),
		autocompletion=remove_completion,
	),
	purge: bool = PURGE,
	debug: bool = DEBUG,
	raw_dpkg: bool = RAW_DPKG,
	download_only: bool = DOWNLOAD_ONLY,
	remove_essential: bool = REMOVE_ESSENTIAL,
	update: bool = UPDATE,
	auto_remove: bool = AUTO_REMOVE,
	fix_broken: bool = FIX_BROKEN,
	assume_yes: bool = ASSUME_YES,
	dpkg_option: list[str] = OPTION,
	verbose: bool = VERBOSE,
	man_help: bool = MAN_HELP,
) -> None:
	"""Remove or Purge packages."""
	command_help("uninstall", "remove", update)
	_remove(pkg_names)


@nala.command("autoremove", help=_("Autoremove packages that are no longer needed."))
@nala.command("autopurge", help=_("Autopurge packages that are no longer needed."))
# pylint: disable=unused-argument,too-many-arguments
def _auto_remove(
	purge: bool = PURGE,
	debug: bool = DEBUG,
	config: bool = CONFIG,
	raw_dpkg: bool = RAW_DPKG,
	download_only: bool = DOWNLOAD_ONLY,
	remove_essential: bool = REMOVE_ESSENTIAL,
	update: bool = UPDATE,
	fix_broken: bool = FIX_BROKEN,
	assume_yes: bool = ASSUME_YES,
	dpkg_option: list[str] = OPTION,
	verbose: bool = VERBOSE,
	man_help: bool = MAN_HELP,
) -> None:
	"""Command for autoremove."""
	sudo_check()
	if config and not arguments.is_purge():
		sys.exit(
			_(
				"{error} {config} must be used with either {autoremove} or {autopurge}."
			).format(
				error=ERROR_PREFIX,
				config=color("--config", "YELLOW"),
				autoremove=color("autoremove --purge", "YELLOW"),
				autopurge=color("autopurge", "YELLOW"),
			)
		)
	cache = setup_cache()
	check_state(cache, nala_pkgs)
	auto_remover(cache, nala_pkgs, config)
	get_changes(cache, nala_pkgs, "remove")


@nala.command(help=_("Show package details."))
@nala.command("info", hidden=True)
# pylint: disable=unused-argument
def show(
	pkg_names: list[str] = typer.Argument(
		...,
		help=_("Package(s) to show"),
		autocompletion=package_completion,
	),
	debug: bool = DEBUG,
	verbose: bool = VERBOSE,
	all_versions: bool = ALL_VERSIONS,
	man_help: bool = MAN_HELP,
) -> None:
	"""Show package details."""
	command_help("info", "show", None)
	cache = Cache()
	not_found: list[str] = []
	pkg_names = cache.glob_filter(pkg_names, show=True)
	pkg_names = cache.virtual_filter(pkg_names)
	additional_records = 0
	for num, pkg_name in enumerate(pkg_names):
		if pkg_name in cache:
			pkg = cache[pkg_name]
			additional_records += show_main(num, pkg)
			continue
		pkg_not_found(pkg_name, cache, not_found)

	if additional_records and not arguments.all_versions:
		additional_notice(additional_records)

	if not_found:
		for error in not_found:
			eprint(error)
		sys.exit(1)


@nala.command(help=_("Search package names and descriptions."))
# pylint: disable=unused-argument,too-many-arguments,too-many-locals
def search(
	regex: str = typer.Argument(
		...,
		help=_("Regex or word to search for"),
		autocompletion=package_completion,
	),
	debug: bool = DEBUG,
	full: bool = FULL,
	names: bool = NAMES,
	installed: bool = INSTALLED,
	nala_installed: bool = NALA_INSTALLED,
	upgradable: bool = UPGRADABLE,
	upgradeable: bool = UPGRADEABLE,
	all_versions: bool = ALL_VERSIONS,
	all_arches: bool = ALL_ARCHES,
	virtual: bool = VIRTUAL,
	verbose: bool = VERBOSE,
	man_help: bool = MAN_HELP,
) -> None:
	"""Search package names and descriptions."""
	cache = Cache()
	found: list[tuple[Package, Version]] = []
	user_installed = (
		get_list(get_history("Nala"), "User-Installed") if nala_installed else []
	)
	if regex == "*":
		regex = ".*"
	try:
		search_pattern = re.compile(regex, re.IGNORECASE)
	except re.error as error:
		sys.exit(
			_(
				"{error} failed regex compilation '{error_msg} at position {position}"
			).format(error=ERROR_PREFIX, error_msg=error.msg, position=error.pos)
		)
	arches = apt_pkg.get_architectures()
	for pkg in cache:
		if nala_installed and pkg.name not in user_installed:
			continue
		if arguments.installed and not pkg.installed:
			continue
		if arguments.upgradable and not pkg.is_upgradable:
			continue
		if arguments.virtual and not cache.is_virtual_package(pkg.name):
			continue
		if arguments.all_arches or pkg.architecture() in arches:
			search_name(pkg, search_pattern, found)

	if not found:
		sys.exit(
			_("{error} {regex} not found.").format(error=ERROR_PREFIX, regex=regex)
		)
	iter_search(found)


@nala.command("list", help=_("List packages based on package names."))
# pylint: disable=unused-argument,too-many-arguments
def list_pkgs(
	pkg_names: Optional[list[str]] = typer.Argument(
		None,
		help=_("Package(s) to list."),
		autocompletion=package_completion,
	),
	debug: bool = DEBUG,
	full: bool = FULL,
	installed: bool = INSTALLED,
	nala_installed: bool = NALA_INSTALLED,
	upgradable: bool = UPGRADABLE,
	upgradeable: bool = UPGRADEABLE,
	all_versions: bool = ALL_VERSIONS,
	virtual: bool = VIRTUAL,
	verbose: bool = VERBOSE,
	man_help: bool = MAN_HELP,
) -> None:
	"""List packages based on package names."""
	cache = Cache()
	user_installed = (
		get_list(get_history("Nala"), "User-Installed") if nala_installed else []
	)

	def _list_gen() -> Generator[
		tuple[Package, Version | tuple[Version, ...]], None, None
	]:
		"""Generate to speed things up."""
		for pkg in cache:
			if nala_installed and pkg.name not in user_installed:
				continue
			if arguments.installed and not pkg.installed:
				continue
			if arguments.upgradable and not pkg.is_upgradable:
				continue
			if arguments.virtual and not cache.is_virtual_package(pkg.name):
				continue
			if pkg_names:
				if pkg.shortname in pkg_names:
					yield (pkg, get_version(pkg, inst_first=True))
				continue
			yield (pkg, get_version(pkg, inst_first=True))

	if not iter_search(_list_gen()):
		sys.exit(_("Nothing was found to list."))


@nala.command(help=_("Clear out the local archive of downloaded package files."))
# pylint: disable=unused-argument
def clean(
	lists: bool = LISTS,
	fetch: bool = FETCH,
	debug: bool = DEBUG,
	verbose: bool = VERBOSE,
	man_help: bool = MAN_HELP,
) -> None:
	"""Clear out the local archive of downloaded package files."""
	sudo_check()
	if lists:
		iter_remove(LISTS_DIR)
		iter_remove(LISTS_PARTIAL_DIR)
		print(_("Package lists have been cleaned"))
		return
	if fetch:
		NALA_SOURCES.unlink(missing_ok=True)
		print(_("Nala sources.list has been cleaned"))
		return
	iter_remove(ARCHIVE_DIR)
	iter_remove(PARTIAL_DIR)
	iter_remove(LISTS_PARTIAL_DIR)
	vprint(
		_("Removing {cache}\nRemoving {src_cache}").format(
			cache=PKGCACHE, src_cache=SRCPKGCACHE
		)
	)

	PKGCACHE.unlink(missing_ok=True)
	SRCPKGCACHE.unlink(missing_ok=True)
	print(_("Cache has been cleaned"))


@nala.command(hidden=True, help=_("I beg, pls moo"))
# pylint: disable=unused-argument
def moo(
	moos: Optional[list[str]] = typer.Argument(None, hidden=True),
	update: bool = typer.Option(None, hidden=True),
) -> None:
	"""I beg, pls moo."""
	print(CAT)
	can_no_moo = _("I can't moo for I'm a cat")
	print(f'{ELLIPSIS}"{can_no_moo}"{ELLIPSIS}')
	if update:
		what_did_you_expect = _("What did you expect to update?")
		print(f'{ELLIPSIS}"{what_did_you_expect}"{ELLIPSIS}')
		return
	if update is not None:
		what_did_you_expect = _("What did you expect no-update to do?")
		print(f'{ELLIPSIS}"{what_did_you_expect}"{ELLIPSIS}')
